h5py/_hl/dataset.py

```diff
@@ -436,7 +441,36 @@ def astype(self, dtype):
         different destination type, e.g.:
 
         >>> double_precision = dataset.astype('f8')[0:100:2]
+
+        Returns
+        -------
+        For dtype='T' (NumPy's native variable-length strings),
+        a writable dataset with the specified dtype.
+
+        For all other dtypes, a read-only view of the dataset
+        with the specified dtype.
         """
+        dtype = numpy.dtype(dtype)
+        if dtype == self.dtype:
+            return self
+
+        if dtype.kind == "T":
+            string_info = h5t.check_string_dtype(self.dtype)
+            if string_info is None:
+                raise TypeError(
+                    f"dset.astype({dtype}) can only be used on datasets with "
+                    "an HDF5 string datatype"
+                )
+            out = self.parent[self.name]
+            assert out.id is not self.id
+            out.id._dtype = dtype
+            return out
+        elif dtype == object and self.dtype.kind == "T":
+            out = self.parent[self.name]
+            assert out.id is not self.id
+            out.id._dtype = None
+            return out
```

Comment:
This seems rather convoluted, and I think it's in service of having writable `.astype('T')` views, which it's not clear to me are actually important. I think if we can get away from switching the dtype of dataset objects, quite a few of the later changes in this file can be avoided as well.

Reply:
Are you saying that `create_dataset(..., dtype="T")` should be disallowed?
Could you write a mockup of the end-user API that you have in mind?